crates/polars-json/src/ndjson/deserialize.rs

```diff
@@ -17,24 +19,45 @@ pub fn deserialize_iter<'a>(
     buf_size: usize,
     count: usize,
 ) -> PolarsResult<ArrayRef> {
-    let mut buf = String::with_capacity(buf_size + count + 2);
+    let mut arr: Vec<Box<dyn Array>> = Vec::new();
+    let mut buf =
+        String::with_capacity(std::cmp::min(buf_size + count + 2, std::u32::MAX as usize));
     buf.push('[');
-    for row in rows {
+
+    fn _deserializer(s: &mut str, data_type: ArrowDataType) -> PolarsResult<Box<dyn Array>> {
+        let slice = unsafe { s.as_bytes_mut() };
+        let out = simd_json::to_borrowed_value(slice)
+            .map_err(|e| PolarsError::ComputeError(format!("json parsing error: '{e}'").into()))?;
+        Ok(if let BorrowedValue::Array(rows) = out {
+            super::super::json::deserialize::_deserialize(&rows, data_type.clone())
+        } else {
+            unreachable!()
+        })
+    }
+    let mut row_iter = rows.peekable();
+
+    while let Some(row) = row_iter.next() {
         buf.push_str(row);
-        buf.push(',')
+        buf.push(',');
+
+        let next_row_length = row_iter.peek().map(|row| row.len()).unwrap_or(0);
+        if buf.len() + next_row_length > std::u32::MAX as usize {
```

Comment:
This doesn't quite work, it should be `>=`. Otherwise you can push to exactly `std::u32::MAX` on one iteration and then on the next iteration do `buf.push(']')` which would put you over `u32::MAX`.

Reply:
@orlp there is a buf.pop() before we do buf.push? Shouldn't that cancel each other out in size?

Reply:
@orlp ah I see!